OSOE-844: Don't include a file called .htmlvalidate.json by default in Lombiq.UITestingToolbox

sarahelsaig opened the pull request:
[OSOE-844](https://lombiq.atlassian.net/browse/OSOE-844)
Fixes #359

[OSOE-844]: https://lombiq.atlassian.net/browse/OSOE-844?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sarahelsaig committed:
Add ``"form-dup-name": "off"` to sample.

sarahelsaig committed:
rename file

sarahelsaig committed:
Update ExecuteTestAfterSetupAsync delegate.

sarahelsaig committed:
Revert "Update ExecuteTestAfterSetupAsync delegate."

This reverts commit ab57abaa469c5fb58b878e4901bf8f80c528948e.

sarahelsaig committed:
Add TestRunTimeout config.

sarahelsaig committed:
Add run timeout handling to UITestBase.ExecuteOrchardCoreTestAsync

sarahelsaig committed:
Update config file name.

sarahelsaig committed:
Add test case.

sarahelsaig committed:
Format code.

sarahelsaig committed:
Add WithRelativeConfigPath.

sarahelsaig committed:
Set the default config path.

sarahelsaig committed:
Fix typos.

sarahelsaig committed:
Code cleanup.

sarahelsaig committed:
Clone TimeoutConfiguration so it won't break when configured per-test.

sarahelsaig committed:
Typo.

sarahelsaig committed:
Fix error propagation.

sarahelsaig committed:
Fix WithRelativeConfigPath.

Piedone commented on Lombiq.Tests.UI/Docs/Configuration.md:
```diff
@@ -59,39 +59,36 @@ Recommendations and notes for such configuration:
 ### HTML validation configuration
 
-If you want to change some HTML validation rules from only a few specific tests, you can create a custom _.htmlvalidate.json_ file (e.g. _TestName.htmlvalidate.json_). For example:
+If you want to change some HTML validation rules from only a few specific tests, you can create a custom _.htmlvalidate.json_ file (e.g. _TestName.htmlvalidate.json_). This should extend the [default.htmlvalidate.json](../default.htmlvalidate.json) file (which is always copied into the build directory) by setting the value of `"extends"` to a relative path pointing to it. For example:
```
After this, also remove the "Make sure to also include the root attribute and set it to true" note below, which is not generally applicable.

```suggestion
If you want to change some HTML validation rules from only a few specific tests, you can create a custom _.htmlvalidate.json_ file (e.g. _TestName.htmlvalidate.json_). This should extend the [default.htmlvalidate.json](../default.htmlvalidate.json) file (which is always copied into the build directory) by setting the value of `"extends"` to a relative path pointing to it and declaring `"root": true`. For example:
```

Piedone commented on Lombiq.Tests.UI/Docs/Configuration.md:
```diff
@@ -59,39 +59,36 @@ Recommendations and notes for such configuration:
+    },
 
-  "root":  true
+    "root":  true
```
There's an extra space after the colon. Please also fix in the files.

Piedone commented on Lombiq.Tests.UI/Docs/Configuration.md:
````diff
@@ -59,39 +59,36 @@ Recommendations and notes for such configuration:
 }
 ```
 
 Then you can change the configuration to use that:
````
```suggestion
You can also create a completely standalone config file too, without any inheritance, but we'd recommend against that. 

You can change the configuration to use the above file as follows:
```

Piedone commented on Lombiq.Tests.UI/Docs/Configuration.md:
```diff
@@ -59,39 +59,36 @@ Recommendations and notes for such configuration:
-                        Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "TestName.htmlvalidate.json")));
+changeConfiguration: configuration => configuration.HtmlValidationConfiguration.HtmlValidationOptions =
+    configuration.HtmlValidationConfiguration.HtmlValidationOptions
+        .CloneWith(validationOptions => validationOptions.ConfigPath =
```
Why do we actually need to clone? I think simply setting the path should be enough.

sarahelsaig replied:
I don't know, it was already like this in the Configuration.md so I assumed it's necessary.
@DemeSzabolcs Do you remember the reason or where it came from? Since you added the docs in #333.

Piedone replied:
FYI I implemented a non-cloning way to set the path under https://github.com/Lombiq/UI-Testing-Toolbox/pull/362 too (will remove it) and nothing bad happened. The config object is unique to the test (ultimately due to `OrchardCoreUITestExecutorConfiguration` being instantiated for each test in `OrchardCoreUITestBase`) so changing its state isn't a problem.

DemeSzabolcs replied:
> I don't know, it was already like this in the Configuration.md so I assumed it's necessary. @DemeSzabolcs Do you remember the reason or where it came from? Since you added the docs in #333.

I did that based on this:
https://github.com/atata-framework/atata-htmlvalidation/issues/8#issuecomment-1864839487

But yeah, `CloneWith` might not needed.

sarahelsaig replied:
Thanks for the context! In this case it's indeed not necessary because the example was using `HtmlValidationOptions.Default`, but we use `new HtmlValidationOptions { ... }` so it's all unique instances.

Piedone commented on Lombiq.Tests.UI.Samples/.htmlvalidate.json:
```diff
@@ -14,6 +14,7 @@
     "wcag/h67": "off",
-    "wcag/h71": "off"
+    "wcag/h71": "off",
+    "form-dup-name": "off"
```
Why didn't you add this into the default config file in the end?

sarahelsaig replied:
omg I added it to the wrong file :scream_cat:

Piedone commented on Lombiq.Tests.UI/Services/HtmlValidationConfiguration.cs:
```diff
@@ -33,6 +34,14 @@ public class HtmlValidationConfiguration
+        // If a consuming project adds a ".htmlvalidate.json" config file then use it, otherwise fall back to the
+        // "default.htmlvalidate.json" which always exists because Lombiq.Tests.UI copies it into the directory during
+        // build.
+        ConfigPath =
+            Path.Combine(AppDomain.CurrentDomain.BaseDirectory, ".htmlvalidate.json") is { } rootConfiguration &&
+            File.Exists(rootConfiguration)
+                ? rootConfiguration
+                : Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "default.htmlvalidate.json"),
```
Easier to understand what's going on.

```suggestion
        ConfigPath =
            Path.Combine(AppDomain.CurrentDomain.BaseDirectory, ".htmlvalidate.json") is { } rootConfiguration && File.Exists(rootConfiguration)
            ? rootConfiguration
            : Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "default.htmlvalidate.json"),
```

Piedone commented on Lombiq.Tests.UI/Docs/Configuration.md:
````diff
@@ -59,39 +59,36 @@ Recommendations and notes for such configuration:
+changeConfiguration: configuration => configuration.HtmlValidationConfiguration.WithRelativeConfigPath("TestName.htmlvalidate.json");
 ```
 
 Make sure to also include the `root` attribute and set it to `true` inside the custom _.htmlvalidate.json_ file and include it in the test project like this:
````
```suggestion
If you want to do this for all tests in the project, just put an _.htmlvalidate.json_ file into the project root and it will be picked up without further configuration.

Make sure to also include the `root` attribute and set it to `true` inside the custom _.htmlvalidate.json_ file and include it in the test project like this:
```

Piedone commented on Lombiq.Tests.UI/Services/HtmlValidationConfiguration.cs:
```diff
@@ -61,6 +70,24 @@ public class HtmlValidationConfiguration
     public Predicate<UITestContext> HtmlValidationAndAssertionOnPageChangeRule { get; set; } =
         EnableOnValidatablePagesHtmlValidationAndAssertionOnPageChangeRule;
 
+    /// <summary>
+    /// Updates the <see cref="HtmlValidationOptions"/>.<see
+    /// cref="Atata.HtmlValidation.HtmlValidationOptions.ConfigPath"/> with a path relative to the <see
+    /// cref="AppDomain.BaseDirectory"/> of the <see cref="AppDomain.CurrentDomain"/> (i.e. the build directory).
+    /// </summary>
```
```suggestion
    /// <summary>
    /// Updates the <see cref="HtmlValidationOptions"/>.<see cref="HtmlValidationOptions.ConfigPath"/> with a path
    /// relative to the <see cref="AppDomain.BaseDirectory"/> of the <see cref="AppDomain.CurrentDomain"/> (i.e. the
    /// build directory).
    /// </summary>
```

Piedone commented on Lombiq.Tests.UI/Services/TimeoutConfiguration.cs:
```diff
@@ -27,13 +35,27 @@ public class TimeoutConfiguration
 
-    public static readonly TimeoutConfiguration Default = new()
+    /// <summary>
+    /// Gets or sets how long an individual test can run to prevent hanging indefinitely. Defaults to 600s.
```
```suggestion
    /// Gets or sets how long an individual test can run to prevent hanging indefinitely. Defaults to 10 minutes.
```

sarahelsaig replied:
I'd prefer to keep the unit of the "Defaults" the same as the unit of the property.

Piedone replied:
The unit of the property is `TimeSpan`, it's no an `int`.

sarahelsaig replied:
The config key is `TimeoutConfiguration:TestRunTimeoutSeconds`.

Piedone replied:
I don't see how that's relevant in the docs of this property. As a developer interacting with this property, you don't need to care about a possible backing config key.

sarahelsaig replied:
If you configure it in one place using seconds but document it in other place using minutes, it's just needless inconsistency. But if you insist...

Piedone commented on Lombiq.Tests.UI.Samples/.htmlvalidate.json:
This should also extend the default.

sarahelsaig committed:
Fix spacing.

sarahelsaig committed:
Organize config files.

sarahelsaig committed:
Update note about root.

sarahelsaig committed:
Merge tag 'asd' into issue/OSOE-844

sarahelsaig committed:
Apply suggestions from code review

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

sarahelsaig committed:
Add suggested paragraph.

sarahelsaig committed:
Merge remote-tracking branch 'origin/issue/OSOE-844' into issue/OSOE-844

sarahelsaig committed:
Fix md.

sarahelsaig committed:
Remove CloneWith.

sarahelsaig committed:
Update Lombiq.Tests.UI/Services/TimeoutConfiguration.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

sarahelsaig committed:
Add missing root option.

Without it html-validate is looking for the "./default.htmlvalidate.json" at the completely wrong location.

sarahelsaig committed:
Merge remote-tracking branch 'origin/issue/OSOE-844' into issue/OSOE-844
